PWA docs - New How-To page about sharing data between apps (#25537)
mdn/content: files/en-us/web/progressive_web_apps/how_to/share_data_between_apps/index.md

To share data, use the {{domxref("navigator.share()")}} method in response to a user action, such as a button click. You can also use the {{domxref("navigator.canShare()")}} method to first validate whether the data you intend to share is, indeed, shareable from the browser that's running your app, and on the current operating system.

### Sharing simple text data

The following example demonstrates how to share a link when a button in the app is clicked:

```js
const button = document.querySelector("#share");

button.addEventListener("click", async () => {
  try {
    await navigator.share({
      title: "Share data between apps",
      text: "An MDN article to learn how to share data between apps",
      url: "https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/How_to/Share_data_between_apps",
    });

    console.log("The URL was successfully shared");
  } catch (err) {
    console.log("The URL could not be shared");
    console.error(err);
  }
});
```

### Sharing files

> While sharing simple text information is well supported, sharing files doesn't work in all browsers. Use the {{domxref("navigator.canShare()")}} method to first validate that the data you want to share is, indeed, shareable.

Author: Oh, then that's something we need to fix in BCD, because sharing url, title, and text definitely works on Firefox (tested on Windows for me).
But, you make a good point that `canShare` should be introduced first.

> While sharing simple text information is well supported, sharing files doesn't work in all browsers. Use the {{domxref("navigator.canShare()")}} method to first validate that the data you want to share is, indeed, shareable.

Author: Ok, I just tested on Firefox release (I use Nightly usually) and the flag is not enabled by default there.
This feature is gated behind `dom.webshare.enabled`, which is true by default in nightly, and false by default in release.

> While sharing simple text information is well supported, sharing files doesn't work in all browsers. Use the {{domxref("navigator.canShare()")}} method to first validate that the data you want to share is, indeed, shareable.

Author: Ok, this should be addressed in the last commit.

It's a good practice to call this method before displaying a button used to share files in your application.

> In the following code example, a file sharing button is hidden if file sharing isn't supported, an event listener is attached and used to share a file otherwise:

Reviewer: ```suggestion
In the following code example, a file sharing button is hidden if file sharing isn't supported; otherwise, an event listener is attached and used to share a file:
```

The `share_target` manifest member allows an installed PWA to be registered, at the operating system level, as a potential target for content shared by other apps. This means that when a user shares some data that's compatible with your PWA, from another app, the operating system will list your PWA alongside other typical share targets like email or messaging apps.

The information you provide with the `share_target` member, in your manifest file, defines which data your app can be a target for, and how the operating system should launch your app when the user selects it as the target.

### Handling simple text data

Here is a web app manifest example using the `share_target` member:

```json
{
  "name": "ChattyBox",
  "start_url": "/",
  "display": "standalone",
  "icons": [
    {
      "src": "images/icon-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "share_target": {
    "action": "/share-handler",
    "method": "GET",
    "params": {
      "text": "description",
      "url": "link"
    }
  }
}
```

> When your app is selected by the user to handle an other app's shared content, it is launched and the shared content is passed to it similar to how {{htmlelement("form")}} elements are submitted.

Reviewer: ```suggestion
When your app is selected by the user to handle an other apps' shared content, it is launched and the shared content is passed to it similar to how {{htmlelement("form")}} elements are submitted.
```

> In the above code example, when the ChattyBox app is selected as a target, it is launched with an HTTP [`GET`](/en-US/docs/Web/HTTP/Methods/GET) request on the `/share-handler` URL with the shared data passed as request parameters named `description` and `link`.

Reviewer: ```suggestion
In the above code example, when the ChattyBox app is selected as a target, it is launched with an HTTP [`GET`](/en-US/docs/Web/HTTP/Methods/GET) request on the `/share-handler` URL, with the shared data passed as request parameters named `description` and `link`.
```

This example shows how to check if the Web Share API is supported and if the data can be shared:

```js
function canBrowserShareData(data) {
  if (!navigator.share || !navigator.canShare) {
    return false;
  }

  return navigator.canShare(data);
}

const sharedDataSample = {
  title: "Some text title",
  text: "More text",
  url: "A url we want to share",
};

if (canBrowserShareData(sharedDataSample)) {
  // Enable the share button in the UI.
  renderAppSharingUI();
} else {
  // We can't share on this browser/operating system.
}
```

### Handling exceptions

The {{domxref("navigator.share()")}} method returns a {{jsxref("Promise")}} that may be rejected in cases such as when the shared data is incorrect, when the user aborts the share operation, or when the data transmission failed.

It's therefore important to catch the promise rejection in order to avoid errors in your app's JavaScript code.

```js
async function shareData(data) {
  try {
    await navigator.share(data);
    // The data was shared successfully.
  } catch (e) {
    // The data could not be shared.

>     console.log(`Error: ${e}`);

Reviewer 2: It always looks odd to me to use `log()` for errors, rather than `error()`.

Author: Good point, will change.

if (canBrowserShareData({ text: "text", url: "https://example.com" })) {
  // Show the button.
  button.style.display = "inline";

  // Listen for click events on the button to share data.
  button.addEventListener("click", async () => {
    try {
      await navigator.share({
        text: "An MDN article to learn how to share data between apps",
        url: "https://developer.mozilla.org/docs/Web/Progressive_web_apps/How_to/Share_data_between_apps",
      });

      console.log("The URL was successfully shared");
    } catch (err) {
      console.log("The URL could not be shared");
      console.error(err);
    }
  });
}
```

### Sharing files

The following code example demonstrates how to share a file when a button in the app is clicked. Similar to the previous example, the `canBrowserShareData` function is used to display the button used to share files only if the browser supports sharing files.

```js
// Retrieve the button from the DOM. The button is hidden for now.
const button = document.querySelector("#share");

// Create some test data with a file, to check if the browser supports
// sharing it.
const testFile = new File(["foo"], "foo.txt", { type: "text/plain" });

> const testSharedData = { files: [testFile] };

Reviewer 2: It seems super ugly that `canShare()` makes us do this. Do you think it is worth hiding this ugliness inside `canBrowserShareData()`? Or maybe having a separate `canBrowserShareFiles()` function to hide it in?

if (canBrowserShareData(testSharedData)) {
  // The browser supports sharing files. Show the button.
  button.style.display = "inline";

  // Listen for clicks on the button and share a file.
  button.addEventListener("click", async () => {
    try {

>       const file = getTheFileToShare();

Reviewer 2: Is it worth having some kind of explanation or even just a hint about what `getTheFileToShare()` might do?

The `share_target` manifest member allows an installed PWA to be registered, at the operating system level, as a potential target for content shared by other apps. This means that when a user shares some data that's compatible with your PWA, from another app, the operating system will list your PWA alongside other typical share targets like email or messaging apps.

The information you provide with the `share_target` member, in your manifest file, defines which data your app can be a target for, and how the operating system should launch your app when the user selects it as the target.

### Handling text data

Here is a web app manifest example using the `share_target` member:

```json
{
  "name": "ChattyBox",
  "start_url": "/",
  "display": "standalone",
  "icons": [
    {
      "src": "images/icon-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "share_target": {
    "action": "/share-handler",
    "method": "GET",
    "params": {
      "text": "description",

>       "url": "link"

Reviewer 3: a screen shot of how this shows up would be fantastic.

Author: I've been trying to find a good example to create an image from, but I couldn't find anything I was happy with so far. Let's keep that in mind for future improvements via another PR I guess.
Part of my problem is that, on Windows, where I'm testing, the share target dialog contains a bunch of things on top of the list of apps, including email recipients that I don't really want to share on MDN.

```

For more information, see the example [Receive share data using GET](/en-US/docs/Web/Manifest/share_target#receiving_share_data_using_get) on the `share_target` web app manifest member page.

### Handling shared files

In the previous example, text data was handled as a `GET` request. However, handling files requires the use of a [`POST`](/en-US/docs/Web/HTTP/Methods/POST) request with a `multipart/form-data` [encoding type](/en-US/docs/Web/API/HTMLFormElement/enctype).

The following code snippet shows how a PWA can be configured to accept different types of shared files:

```json
{
  "name": "ChattyBox",
  "start_url": "/",
  "display": "standalone",
  "icons": [
    {
      "src": "images/icon-256.png",
      "sizes": "256x256",
      "type": "image/png"
    }
  ],
  "share_target": {
    "action": "/share-file-handler",
    "method": "POST",
    "enctype": "multipart/form-data",
    "params": {
      "files": [
        {
          "name": "textFiles",
          "accept": ["text/plain", ".txt"]
        },
        {
          "name": "htmlFiles",
          "accept": ["text/html", ".html"]
        },
        {
          "name": "images",

>           "accept": ["image/jpeg", "image/png", "image/webp", "image/gif"]

Reviewer 3: Do you want to add one `"accept": ["image/*"]` or similar to show that every type needs to be explicitly written

        }
      ]
    }
  }
}
```

As this example shows, each file object in the `files` property must have a `name` property and an `accept` property. The `accept` property must specify the accepted [MIME types](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) or file extensions.

When the app is selected by the user to handle a shared file (or files), the app is launched with a `POST` request at the `/share-file-handler` URL, with encoded form data.

> Because this is a `POST` request, your app's main JavaScript code can't access the form data directly. You can handle the submitted files in your server-side code, by receiving them at the `/share-file-handler` URL end point. However, for a better user experience that works offline, you can handle the files in your service worker code with a {{domxref("FetchEvent", "fetch")}} event listener, as shown here:

Reviewer 2: I think https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/fetch_event would be a better link than https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent here.

> Because this is a `POST` request, your app's main JavaScript code can't access the form data directly. You can handle the submitted files in your server-side code, by receiving them at the `/share-file-handler` URL end point. However, for a better user experience that works offline, you can handle the files in your service worker code with a {{domxref("FetchEvent", "fetch")}} event listener, as shown here:

Reviewer 3: Not sure if you want to mention GET, but if you do, here is a suggested edit:

```suggestion
Because this is a `POST` request, your app's main JavaScript code can't access the form data directly. You can handle the submitted files in your server-side code by receiving them at the `/share-file-handler` URL endpoint. If you need to access the form data client-side, use `GET`, realizing that the information submitted is less secure.

For a better user experience that works offline, you can handle the files in your service worker code with a {{domxref("FetchEvent", "fetch")}} event listener, as shown here:
```

Author: When sharing files, you can't use GET. Using POST and `multipart/form-data` is required. So I don't think we should mention GET here.

```js
// service-worker.js

self.addEventListener("fetch", (event) => {
  // Only use this event listener for POST requests sent to /share-file-handler.
  const url = new URL(event.request.url);
  if (
    event.request.method !== "POST" ||
    url.pathname !== "/share-file-handler"
  ) {
    return;
  }

  event.respondWith(
    (async () => {
      // Get the data from the submitted form.
      const formData = await event.request.formData();

      // Get the submitted files.
      const textFiles = formData.getAll("textFiles");
      const htmlFiles = formData.getAll("htmlFiles");
      const imageFiles = formData.getAll("images");

      // Send the files to the frontend app.
      sendFilesToFrontend(textFiles, htmlFiles, imageFiles);

      // Redirect the user to a URL that shows the imported files.
      return Response.redirect("/display-new-files", 303);
    })()
  );
});
```

> In this code example, the shared files are extracted from the form data, and the user is redirected to a different page. It's up to you, using the code in your service worker, to handle the extracted files as you want. For example, you can send them to your app's main JavaScript code using the {{domxref("Worker.postMessage()")}} method, or store them in an {{domxref("IndexedDB")}} database which can be accessed both by your service worker and app's main JavaScript code.

Reviewer 3: ```suggestion
In this code example, the shared files are extracted from the form data and the user is redirected to a different page. It's up to you, using the code in your service worker, to handle the extracted files as you want. For example, you can send them to your app's main JavaScript code using the {{domxref("Worker.postMessage()")}} method or store them in an {{domxref("IndexedDB")}} database which can be accessed both by your service worker and app's main JavaScript code.
```

> For more information, see the example [Receiving shared files](/en-US/docs/Web/Manifest/share_target#receiving_shared_files) on the the `share_target` web app manifest member page.

Reviewer 3: ```suggestion
For more information, see the [receiving shared files](/en-US/docs/Web/Manifest/share_target#receiving_shared_files) example on the `share_target` web app manifest member page.
```
